test/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test/util/solo_test.dart invariant_boolean`
+
+int bar = 1;
+
+int baz = 2;
+
+int foo = 0;
+bool setting = null;
+
+void bad() {
+  if ((foo == bar && someComputation()) || (foo != bar && otherComputation())) {} // OK
+  if (foo > bar || foo == bar) {} // OK
+  if (foo > bar && someComputation() || foo == bar) {} // OK
+  if (foo > bar || foo == bar && someComputation()) {} // OK
+  if (foo < bar - baz || foo > bar) {} // OK
+  if (setting == true) {} // OK
+  if (setting != true) {} // OK
+  if (foo == bar || foo != bar) {} // LINT
```

Comment:
I notice that all of these tests are based on top-level variables, but I have concerns about false positives when the identifiers reference getters, which can return different values on each invocation. I can't think of a realistic example that would involve the same code patterns, but the spirit of the rule is that the values of the expressions are statically knowable (in some abstract sense) and hence we can reason about them. Getters violate that by being able to return different values on each call. For example, what if `foo` were a getter that returned and incremented a counter on each invocation?